src/unused_label.rs

```diff
@@ -0,0 +1,81 @@
+use rustc::lint::*;
+use rustc_front::hir;
+use rustc_front::intravisit::FnKind;
+use std::collections::{HashMap, HashSet};
+use syntax::ast;
+use syntax::codemap::Span;
+use syntax::parse::token::InternedString;
+use utils::span_lint;
+
+/// **What it does:** This lint checks for unused labels.
+///
+/// **Why is this bad?** Maybe the label should be used in which case there is an error in the
+/// code or it should be removed.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust,ignore
+/// fn unused_label() {
+///     'label: for i in 1..2 {
+///         if i > 4 { continue }
+///     }
+/// ```
+declare_lint! {
+    pub UNUSED_LABEL,
+    Warn,
+    "unused label"
+}
+
+pub struct UnusedLabel {
+    declared_labels: HashMap<InternedString, Span>,
+    used_labels: HashSet<InternedString>,
+}
+
+impl ::std::default::Default for UnusedLabel {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl UnusedLabel {
+    pub fn new() -> UnusedLabel {
+        UnusedLabel {
+            declared_labels: HashMap::new(),
+            used_labels: HashSet::new(),
+        }
+    }
+}
+
+impl LintPass for UnusedLabel {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNUSED_LABEL)
+    }
+}
+
+impl LateLintPass for UnusedLabel {
+    fn check_fn(&mut self, _: &LateContext, _: FnKind, _: &hir::FnDecl, _: &hir::Block, _: Span, _: ast::NodeId) {
+        self.declared_labels.clear();
+        self.used_labels.clear();
+    }
+
+    fn check_expr(&mut self, _: &LateContext, expr: &hir::Expr) {
+        match expr.node {
+            hir::ExprBreak(Some(label)) | hir::ExprAgain(Some(label)) => {
+                self.used_labels.insert(label.node.name.as_str());
```

Comment:
you could just remove the appropriate label from `declared_labels`, then you just need to iterate over the hash-map at the end instead of indexing it from the values in the hashset